clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp

```diff
@@ -683,6 +683,30 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
 
     return Builder.CreateInsertElement(I0, A, 1);
   }
+  case AMDGPU::BI__builtin_amdgcn_image_load_2d_f32_i32: {
+    llvm::Type *RetTy = llvm::Type::getFloatTy(Builder.getContext());
+    llvm::Type *IntTy = llvm::IntegerType::get(Builder.getContext(), 32u);
+
+    llvm::Value *imm0 = llvm::ConstantInt::get(IntTy, 1);
```

Comment:
Is is the dmask? 

For images, can we get away with only having the builtins return the 4 element case, and rely on demanded elements to shrink the dmask and return type? We do that optimization for buffers but I don't see it for images

Reply:
> For images, can we get away with only having the builtins return the 4 element case, and rely on demanded elements to shrink the dmask and return type? We do that optimization for buffers but I don't see it for images

It's here: https://github.com/llvm/llvm-project/blob/22576e2ccec60af6d27d8fd95ad3ca721b914815/llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp#L1715